src/fetchers/onchain.js

```diff
@@ -0,0 +1,199 @@
+};
+
+setInterval(() => {
+  // Every 10 minutes, we refresh the web3 instances to bring old memory out of scope to avoid memory leaks
+  Web3Instances = [];
+  for (let i = 0; i < Web3InstanceMax; i++) {
+    Web3Instances.push(new web3(process.env.MAINNET_RPC_URL));
+  }
+}, 60 * 1000);
+
+const encodeTokenERC721 = (token) => {
+  return {
+    id: token.requestId,
+    encodedTokenID: Web3Local.eth.abi.encodeFunctionCall(
+      {
+        name: "tokenURI",
+        type: "function",
+        inputs: [
+          {
+            type: "uint256",
+            name: "tokenId",
+          },
+        ],
+      },
+      [token.tokenId]
+    ),
+    contract: token.contract,
+  };
+};
+
+const encodeTokenERC1155 = (token) => {
+  return {
+    id: token.requestId,
+    encodedTokenID: Web3Local.eth.abi.encodeFunctionCall(
+      {
+        name: "uri",
+        type: "function",
+        inputs: [
+          {
+            type: "uint256",
+            name: "tokenId",
+          },
+        ],
+      },
+      [token.tokenId]
+    ),
+    contract: token.contract,
+  };
+};
+
+const createBatch = (encodedTokens) => {
+  return encodedTokens.map((token) => {
+    return {
+      jsonrpc: "2.0",
+      id: token.id,
+      method: "eth_call",
+      params: [
+        {
+          data: token.encodedTokenID,
+          to: token.contract,
+        },
+        "latest",
+      ],
+    };
+  });
+};
+
+const sendBatch = async (encodedTokens, RPC_URL) => {
+  try {
+    const response = await fetch(RPC_URL, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(createBatch(encodedTokens)),
+      timeout: FETCH_TIMEOUT,
+      // TODO: add proxy support to avoid rate limiting
+      // agent:
+    });
+    const json = await response.json();
+    return [json, null];
+  } catch (e) {
+    return [null, e.message];
+  }
+};
+
+const getTokenMetadataFromURI = async (uri) => {
+  try {
+    if (uri.includes("ipfs://")) {
+      uri = uri.replace("ipfs://", "https://ipfs.io/ipfs/");
+    }
+    if (!uri.includes("http")) {
+      return [null, "Invalid URL"];
+    }
+
+    const response = await fetch(uri, {
+      method: "GET",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      timeout: FETCH_TIMEOUT,
+      // TODO: add proxy support to avoid rate limiting
+      // agent:
+    });
+
+    if (!response.ok) {
+      return [null, response.status];
+    }
+
+    const json = await response.json();
+    return [json, null];
+  } catch (e) {
+    return [null, e.message];
+  }
+};
+
+export const fetchTokens = async (chainId, tokens, standard = "ERC721") => {
+  // TODO: Add support for other chains via RPC_URL
+  if (tokens.length === 0) return [];
+  if (!Array.isArray(tokens)) tokens = [tokens];
+  //   if (tokens.length > 20) throw new Error("Too many tokenIds (max 20)");
+  if (chainId !== 1) throw new Error("Only mainnet is supported");
```

Comment:
Can you explain this limitation? The goal is to maybe utilize the on chain on chains that are not supported by opensea / simplehash.

Reply:
i think i had this in because i was going to ask for RPCs for the other chains, ill go in and add that logic